crates/matrix-sdk/src/event_cache/linked_chunk.rs

```diff
@@ -26,12 +28,31 @@ use std::{
 };
 
 /// Errors of [`LinkedChunk`].
-#[derive(Debug)]
+#[derive(thiserror::Error, Debug)]
 pub enum LinkedChunkError {
+    #[error("The chunk identifier is invalid: `{identifier:?}`")]
     InvalidChunkIdentifier { identifier: ChunkIdentifier },
+
+    #[error("The chunk is a gap: `{identifier:?}`")]
     ChunkIsAGap { identifier: ChunkIdentifier },
+
+    #[error("The chunk is an item: `{identifier:?}`")]
     ChunkIsItems { identifier: ChunkIdentifier },
+
+    #[error("The item index is invalid: `{index}`")]
     InvalidItemIndex { index: usize },
+
+    #[error("An error from the listener: `{0}`")]
+    Listener(Box<dyn Error + Send + Sync>),
```

Comment:
Can you explain what's the requirement for this?

When I consider the use case of a listener to store things to disk storage, I don't think a disk error should lead to a linked chunk error. Since some linked chunk operations may cause multiple listener calls (because we insert a new chunk, then insert things into the new chunk), that means we could have partially correct state, and no way to recover from that.

I think a more robust design would be to not allow errors inside the listener, and let the disk storage listener fail in another way. Alternatively, the disk storage listener could store all the updates to the storage in one hand, and then apply them to the storage later, using a command pattern. It's a bit more involved, but it would better separate the concerns of "the in-memory linked chunks are corrupted" vs "the on-disk storage is corrupted".

What do you think?

Reply:
No more present!